From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: quh4gko8 <88831734+quh4gko8@users.noreply.github.com>
Date: Mon, 27 Apr 2026 12:45:57 +0000
Subject: [PATCH] [bugfix] Avoid unneeded calls to clipboard primary contents
 at URL bar

The recent change to use async GetAllAvailableFormats upstream, as seen
at [1] causes ClipboardManager#getPrimaryClip() calls through the
clipboard wrapper upstream, specifically ClipboardImpl#getFilenames().

Stub out calls to clipboard wrapper that checks for primary clip
directly when its result is not used for clipboard format checks. One
such case is on autocomplete clipboard checks at the url bar.

[1] https://chromium-review.googlesource.com/c/chromium/src/+/7631097
---
 .../suggestions/AutocompleteController.java   |  3 +++
 .../src/org/chromium/ui/base/Clipboard.java   |  3 +++
 .../org/chromium/ui/base/ClipboardImpl.java   | 19 +++++++++++++++++++
 3 files changed, 25 insertions(+)

diff --git a/chrome/browser/ui/android/omnibox/java/src/org/chromium/chrome/browser/omnibox/suggestions/AutocompleteController.java b/chrome/browser/ui/android/omnibox/java/src/org/chromium/chrome/browser/omnibox/suggestions/AutocompleteController.java
index 7953cdbe0f872..7550207c88868 100644
--- a/chrome/browser/ui/android/omnibox/java/src/org/chromium/chrome/browser/omnibox/suggestions/AutocompleteController.java
+++ b/chrome/browser/ui/android/omnibox/java/src/org/chromium/chrome/browser/omnibox/suggestions/AutocompleteController.java
@@ -117,6 +117,7 @@ public class AutocompleteController {
             boolean preventInlineAutocomplete) {
         if (mNativeController == 0) return;
 
+        org.chromium.ui.base.Clipboard.getInstance().setStubHasFilenamesCallOneTime(true);
         AutocompleteControllerJni.get()
                 .start(
                         mNativeController,
@@ -178,6 +179,7 @@ public class AutocompleteController {
      */
     public @Nullable AutocompleteMatch classify(String text) {
         if (mNativeController == 0) return null;
+        org.chromium.ui.base.Clipboard.getInstance().setStubHasFilenamesCallOneTime(true);
         return AutocompleteControllerJni.get().classify(mNativeController, text);
     }
 
@@ -190,6 +192,7 @@ public class AutocompleteController {
     public void startZeroSuggest(@Nullable WebContents webContents, AutocompleteInput input) {
         if (mNativeController == 0) return;
 
+        org.chromium.ui.base.Clipboard.getInstance().setStubHasFilenamesCallOneTime(true);
         AutocompleteControllerJni.get()
                 .onOmniboxFocused(
                         mNativeController,
diff --git a/ui/android/java/src/org/chromium/ui/base/Clipboard.java b/ui/android/java/src/org/chromium/ui/base/Clipboard.java
index 3f542c5f5151a..302522687151e 100644
--- a/ui/android/java/src/org/chromium/ui/base/Clipboard.java
+++ b/ui/android/java/src/org/chromium/ui/base/Clipboard.java
@@ -254,6 +254,9 @@ public class Clipboard {
         return false;
     }
 
+    public void setStubHasFilenamesCallOneTime(boolean shouldStub) {
+    }
+
     /**
      * Emulates the behavior of the now-deprecated {@link
      * android.text.ClipboardManager#setText(CharSequence)}, setting the clipboard's current primary
diff --git a/ui/android/java/src/org/chromium/ui/base/ClipboardImpl.java b/ui/android/java/src/org/chromium/ui/base/ClipboardImpl.java
index a767173284ffe..009fc29de2a6b 100644
--- a/ui/android/java/src/org/chromium/ui/base/ClipboardImpl.java
+++ b/ui/android/java/src/org/chromium/ui/base/ClipboardImpl.java
@@ -415,8 +415,27 @@ public class ClipboardImpl extends Clipboard
         return uris.toArray(new String[][] {});
     }
 
+    private final java.util.concurrent.atomic.AtomicBoolean mStubHasFilenamesCallOneTime =
+            new java.util.concurrent.atomic.AtomicBoolean(false);
+
+    @Override
+    public void setStubHasFilenamesCallOneTime(boolean shouldStub) {
+        mStubHasFilenamesCallOneTime.set(shouldStub);
+    }
+
     @Override
     public boolean hasFilenames() {
+        final String LOG_TAG = "ClipboardImpl_Debug";
+        if (mStubHasFilenamesCallOneTime.getAndSet(false)) {
+            if (org.chromium.base.Log.isLoggable(LOG_TAG, org.chromium.base.Log.DEBUG)) {
+                org.chromium.base.Log.i(LOG_TAG, "stubbing hasFilenames()", new Throwable());
+            }
+
+            return super.hasFilenames();
+        } else if (org.chromium.base.Log.isLoggable(LOG_TAG, org.chromium.base.Log.DEBUG)) {
+            org.chromium.base.Log.i(LOG_TAG, "not stubbing hasFilenames()", new Throwable());
+        }
+
         // getPrimaryClip() has been observed to throw unexpected exceptions for some devices (see
         // crbug/654802 and b/31501780)
         try {
